janus_collector/src/lib.rs

```diff
@@ -344,16 +345,39 @@ where
                 request.send().await
             },
         )
-        .await
-        .map_err(|res| match res {
-            Ok(response) => Error::Http(response.status()),
-            Err(error) => Error::HttpClient(error),
-        })?;
-
-        let status = response.status();
-        if status != StatusCode::SEE_OTHER {
-            return Err(Error::Http(status));
-        }
+        .await;
+
+        let response = match response_res {
+            // Successful response or unretryable error status code:
+            Ok(response) => {
+                let status = response.status();
+                if status == StatusCode::SEE_OTHER {
+                    response
+                } else if status.is_client_error() || status.is_server_error() {
+                    if let Ok(mut problem_details) = response.json::<HttpApiProblem>().await {
+                        problem_details.status = Some(status);
+                        return Err(Error::Http(problem_details));
+                    } else {
+                        return Err(Error::Http(HttpApiProblem::new(status)));
+                    }
+                } else {
+                    return Err(Error::Http(HttpApiProblem::new(status)));
+                }
+            }
+            // Retryable error status code, but ran out of retries:
+            Err(Ok(response)) => {
```

Comment:
nit: these two match arms are very similar (but not identical), and similar logic appears below as well. Could we introduce a helper function (or something) to avoid repeating the process-body-into-problem-document logic?